plugins/inputs/marklogic/marklogic.go

```diff
@@ -0,0 +1,231 @@
+			Online         MlPointBool `json:"online"`
+			LoadProperties struct {
+				TotalLoad MlPointFloat `json:"total-load"`
+			} `json:"load-properties"`
+			RateProperties struct {
+				TotalRate MlPointFloat `json:"total-rate"`
+			} `json:"rate-properties"`
+			StatusDetail struct {
+				Cpus                   MlPointInt `json:"cpus"`
+				Cores                  MlPointInt `json:"cores"`
+				TotalCPUStatUser       float64    `json:"total-cpu-stat-user"`
+				TotalCPUStatSystem     float64    `json:"total-cpu-stat-system"`
+				TotalCPUStatIdle       float64    `json:"total-cpu-stat-idle"`
+				TotalCPUStatIowait     float64    `json:"total-cpu-stat-iowait"`
+				MemoryProcessSize      MlPointInt `json:"memory-process-size"`
+				MemoryProcessRss       MlPointInt `json:"memory-process-rss"`
+				MemorySystemTotal      MlPointInt `json:"memory-system-total"`
+				MemorySystemFree       MlPointInt `json:"memory-system-free"`
+				MemorySize             MlPointInt `json:"memory-size"`
+				HostSize               MlPointInt `json:"host-size"`
+				DataDirSpace           MlPointInt `json:"data-dir-space"`
+				QueryReadBytes         MlPointInt `json:"query-read-bytes"`
+				QueryReadLoad          MlPointInt `json:"query-read-load"`
+				HTTPServerReceiveBytes MlPointInt `json:"http-server-receive-bytes"`
+				HTTPServerSendBytes    MlPointInt `json:"http-server-send-bytes"`
+			} `json:"status-detail"`
+		} `json:"status-properties"`
+	} `json:"host-status"`
+}
+
+// Description of plugin returned
+func (c *Marklogic) Description() string {
+	return "Retrives information on a specific host in a MarkLogic Cluster"
+}
+
+var sampleConfig = `
+  ## Base URL of the MarkLogic HTTP Server.
+  url = "http://localhost:8002"
+
+  ## List of specific hostnames to retrieve information. At least (1) required.
+  # hosts = ["hostname1", "hostname2"]
+
+  ## Using HTTP Digest Authentication. Management API requires 'manage-user' role privileges
+  # username = "myuser"
+  # password = "mypassword"
+
+  ## Optional TLS Config
+  # tls_ca = "/etc/telegraf/ca.pem"
+  # tls_cert = "/etc/telegraf/cert.pem"
+  # tls_key = "/etc/telegraf/key.pem"
+  ## Use TLS but skip chain & host verification
+  # insecure_skip_verify = false
+`
+
+// SampleConfig to gather stats from localhost, default port.
+func (c *Marklogic) SampleConfig() string {
+	return sampleConfig
+}
+
+// Gather metrics from HTTP Server.
+func (c *Marklogic) Gather(accumulator telegraf.Accumulator) error {
+	var wg sync.WaitGroup
+	var url string
+
+	if c.client == nil {
+		client, err := c.createHttpClient()
+
+		if err != nil {
+			return err
+		}
+		c.client = client
+	}
+
+	if len(c.URL) == 0 {
+		c.URL = string("http://localhost:8002")
+	}
+
+	// Range over all ML hostnames configured in fiter. Returns early in case of any error.
+	for _, u := range c.Hosts {
+		wg.Add(1)
+		go func(host string) {
+			defer wg.Done()
+
+			url = string(c.URL + statsPath + host + viewFormat)
```

Comment:
Use `url.URL` here to avoid malformed URLs.  Something like this but handling errors:
```
base, err := url.Parse(c.URL)
addr := base.ResolveReference(path.Join(statsPath, host))
addr.RawQuery(viewFormat) # but first remove the leading "?"
u = addr.String()
```

Consider doing this only once and saving the list of URLs on the struct.  If you like you could use a Init() function https://github.com/influxdata/telegraf/blob/master/input.go#L9

Reply:
Great suggestion! Your like Obi-Wan Kenobi.  Added lines to do the url parsing. Also using the Init() function to parse and save all urls to the struct just once.

```
func (c *Marklogic) Init()  error {}. . .
```